### PR TITLE
Add support for "Training RNNs as fast as CNNs"

### PR DESCRIPTION
**Pull request by JianyuZhan**
This PR addresses Issue https://github.com/OpenNMT/OpenNMT-py/issues/251, trying to incorporate the author's implementation into our project. I am doing this by adding a new `-rnn_type`: **SRU** in `RNNEncoder` and `StdRNNDecoder`, not to create a new `SRUEncoder` and `SRUDecoder`. I think this makes much more sense - `SRU` is just a variant of `RNN`, the same as `LSTM` and `GRU`. Thanks to our modular structure, I am trying to keep this less intrusive to the code structure.  Hopefully sometime in the future when pytorch community supports this RNN variant, we could easily migrate to use it.

**Patch list**

This PR has 2 patches.

**1.** 97929d5 **RNNEncoder: Add a command-line option to disable pack_padded_sequence**

This is a preparing patch.  Since this SRU doesn't support `PackedSequence`,  rather to create a almost the same new `NoPackedSeqSupportRNNEncoder`, I just incorporates an option `no_pack_padded_seq`, which defaults to be `False`.  And when it is set, the `lengths` argument in encoder's `forward()` is just ignored and no `pack/unpack` behavior is taken.  When `SRU` is used, it is mandated to also set this option, or the below error will be spewed:

```
Exception: 'SRU' doesn't support PackedSequence, please set option: -no_pack_padded_seq!
```

**2.** e9a07db **Introduce a rnn_type: SRU**

This patch introduces the SRU implementation, with slight modification. The whole code locates in
`onmt/modules/SRU.py`.

This SRU implementation does the cuda level optimization by itself and thus requires the `cupy` and `pynvrtc` packages installed, also requires the cuda library path is set in the `LD_LIBRARY_PATH` environment settings.

We don't mandate these requirements to not upset users who don't use `SRU`.  So a conditional requirement is checked, and I incorporates a function called `check_sru_requirement()` in `opts.py`, and do checks in these places:

1.  **In the unittest**.  The test of `-rnn_type SRU` is only performed when `check_sru_requirement()` is satisfied.
2. **In the `onmt/modules/__init__.py`.**  `SRU` is only exported to other modules when `check_sru_requirement()` is satisfied.
3. **In the command line parsing**.  When user explicitly requires `-rnn_type SRU`,  `check_sru_requirement()`  is also called by using `argparse.Action`. And since this is user's purpose,
we abort if these requirements are not satisfied and relevant error is printed:

```
Exception: Using 'SRU' requires 'cupy' and 'pynvrtc' python packages installed.
```

**Test  Done**
1. unittest is performed **w/** and **w/o** these requirements, passed.
2. `python train.py` is run with `-rnn_type SRU`,  **w/**  and **w/o**  `-no_pack_padded_seq`, as designed.

One interesting observation is that the training perplexity rockets really high at first(Train perplexity: 1.34152e+10 in Epoch 2),  but latter when learning rate drops, it drops to 2000 in Epoch 8, and stops at 1785.05 in Epoch 13, the final one).

### CODE REVIEW

File: onmt/Models.py [hunk 1]
```diff
@@ -53,34 +53,54 @@ def forward(self, input, lengths=None, hidden=None):
 class RNNEncoder(EncoderBase):
     """ The standard RNN encoder. """
     def __init__(self, rnn_type, bidirectional, num_layers,
-                 hidden_size, dropout, embeddings):
+                 hidden_size, no_pack_padded_seq,
```
Comment (srush) [inline onmt/Models.py]: This doesn't need to be an option that users know about. Just have it be set automatically if they pass SRU

File: opts.py [hunk 2]
```diff
@@ -57,9 +60,28 @@ def model_opts(parser):
                         help="""Feed the context vector at each time step as
                         additional input (via concatenation with the word
                         embeddings) to the decoder.""")
+
+    class CheckSRU(argparse.Action):
```
Comment (srush) [inline opts.py]: Put this in the SRU module, don't clutter opts.

File: opts.py [hunk 3]
```diff
@@ -218,6 +240,37 @@ def add_md_help_argument(parser):
                         help='print Markdown-formatted help text and exit.')
 
 
+# Our current SRU version implements its own cuda-level optimization,
+# so it needs the `cupy` and `pynvrtc` package. It also needs the cuda
+# library path set, e.g.: 'export LD_LIBRARY_PATH=/usr/local/cuda/lib64'.
+def check_sru_requirement(abort=False):
```
Comment (srush) [inline opts.py]: Also in SRU module.

**Comment by JianyuZhan** [comment]
**v1->v2**

added a new commit, addressing the review feedback in v1:

d771f2f SRU: move sru requirement checks and hide PackedSeqeunce check from user
